### PR TITLE
Cross cluster status updates

### PR DESCRIPTION
**Pull request by desagar**
# Description

Send status updates from the managed cluster to the admin cluster for all MultiClusterXXX resources

Fixes VZ-2265

# Checklist 

As the author of this PR, I have:

- [x] Checked that I included or updated copyright and license notices in all files that I altered
- [x] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: application-operator/apis/clusters/v1alpha1/multiclusterapplicationconfiguration_types.go [hunk 1]
```diff
@@ -50,3 +50,8 @@ type MultiClusterApplicationConfigurationList struct {
 func init() {
 	SchemeBuilder.Register(&MultiClusterApplicationConfiguration{}, &MultiClusterApplicationConfigurationList{})
 }
+
+// GetStatus returns the MultiClusterResourceStatus of this resource
+func (in *MultiClusterApplicationConfiguration) GetStatus() MultiClusterResourceStatus {
```
Comment (pfmackin) [inline application-operator/apis/clusters/v1alpha1/multiclusterapplicationconfiguration_types.go:55]: why is this needed?   Normally you would just access the Status field directly

File: application-operator/mcagent/mcagent_common.go [hunk 2]
```diff
@@ -35,3 +40,32 @@ func (s *Syncer) isThisCluster(placement clustersv1alpha1.Placement) bool {
 	}
 	return false
 }
+
+// processStatusUpdates monitors the StatusUpdateChannel for any
+// received messages and processes a batch of them
+func (s *Syncer) processStatusUpdates() {
```
Comment (pfmackin) [inline application-operator/mcagent/mcagent_common.go:46]: I think we want only the last status update for a given resource.  If that is the case, then i suggest we use a map with the resource namespacedName as a key and process all message (not limited to batch size).  Then we just do max one update per resource.

File: application-operator/apis/clusters/v1alpha1/multiclusterapplicationconfiguration_types.go (hunk 1 above)
Comment (desagar) [reply application-operator/apis/clusters/v1alpha1/multiclusterapplicationconfiguration_types.go:55]: This is to satisfy an interface so that I can use a single method to do the UpdateStatus in cluster utils, which accepts an interface type

File: application-operator/controllers/clusters/cluster_utils.go [hunk 3]
```diff
@@ -248,3 +268,37 @@ func fetchClusterSecret(ctx context.Context, rdr client.Reader, clusterSecret *c
...
+// the status update
+func UpdateStatus(resource MultiClusterResource, mcStatus *clustersv1alpha1.MultiClusterResourceStatus, placement clustersv1alpha1.Placement, newCondition clustersv1alpha1.Condition, clusterName string, agentChannel chan StatusUpdateMessage, updateFunc func() error) (controllerruntime.Result, error) {
+
+	clusterLevelStatus := CreateClusterLevelStatus(newCondition, clusterName)
+
+	if StatusNeedsUpdate(*mcStatus, newCondition, clusterLevelStatus) {
+		mcStatus.Conditions = append(mcStatus.Conditions, newCondition)
+		UpdateClusterLevelStatus(mcStatus, clusterLevelStatus)
+		mcStatus.State = ComputeEffectiveState(*mcStatus, placement)
+		fmt.Printf("UpdateStatus: Calling update status func for %s/%s on cluster %s = %s\n", resource.GetNamespace(), resource.GetName(),
```
[the remark this answers is not in the document]
Comment (desagar) [reply application-operator/controllers/clusters/cluster_utils.go]: removed

File: application-operator/mcagent/mcagent_common.go (hunk 2 above)
Comment (desagar) [reply application-operator/mcagent/mcagent_common.go:50]: I decided to drop this optimization since there was a lot else to do. I'm processing small batches at the time so I expect it's not too bad.

File: application-operator/controllers/clusters/cluster_utils.go [hunk 4]
```diff
@@ -248,3 +268,46 @@ func fetchClusterSecret(ctx context.Context, rdr client.Reader, clusterSecret *c
...
+	if StatusNeedsUpdate(*mcStatus, newCondition, clusterLevelStatus) {
+		mcStatus.Conditions = append(mcStatus.Conditions, newCondition)
+		SetClusterLevelStatus(mcStatus, clusterLevelStatus)
+		mcStatus.State = ComputeEffectiveState(*mcStatus, placement)
+		err := updateFunc()
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		if agentChannel != nil {
+			// put the status update itself on the agent channel. TODO may need to do a deep copy for the cross-thread dereferencing to work
```
Comment (mgianatagh) [inline application-operator/controllers/clusters/cluster_utils.go:288]: will there be a Jira created for this todo?

Comment (mgianatagh) [reply application-operator/controllers/clusters/cluster_utils.go:288]: Or should we just do the deep copy now?  Is it more than a call to an existing function?

File: application-operator/constants/constants.go [hunk 5]
```diff
@@ -56,3 +56,11 @@ const LabelIstioInjection = "istio-injection"
 
 // LabelIstioInjectionDefault - default value for LabelIstioInjection
 const LabelIstioInjectionDefault = "enabled"
+
+// StatusUpdateChannelBufferSize - the number of status update messages that will be buffered
+// by the agent channel before controllers trying to send more status updates will start blocking
+const StatusUpdateChannelBufferSize = 10
```
Comment (desagar) [inline application-operator/constants/constants.go:62]: update to 50 for now

File: application-operator/mcagent/mcagent_common.go [hunk 6]
```diff
@@ -35,3 +44,57 @@ func (s *Syncer) isThisCluster(placement clustersv1alpha1.Placement) bool {
...
+		// Use a select with default so as to not block on the channel if there are no updates
+		select {
+		case msg := <-s.StatusUpdateChannel:
+			s.Log.V(2).Info(fmt.Sprintf("processStatusUpdates: Received status update %s with condition type %s for %s/%s from cluster %s",
+				msg.NewClusterStatus.State, msg.NewCondition.Type, msg.Resource.GetNamespace(), msg.Resource.GetName(), msg.NewClusterStatus.Name))
+			err := s.performAdminStatusUpdate(msg)
+			if err != nil {
+				if errors.IsConflict(err) {
+					// Retry on conflict - put the message back on the channel
+					s.StatusUpdateChannel <- msg
```
Comment (desagar) [inline application-operator/mcagent/mcagent_common.go:61]: this is a deadlock risk

Comment (desagar) [reply application-operator/mcagent/mcagent_common.go:61]: to prevent a deadlock, can use a select with send

File: application-operator/controllers/clusters/cluster_utils.go (hunk 4 above)
Comment (desagar) [reply application-operator/controllers/clusters/cluster_utils.go]: this TODO is not needed.

File: application-operator/mcagent/mcagent_common.go (hunk 6 above)
Comment (desagar) [reply application-operator/mcagent/mcagent_common.go]: implemented a fix for the possible deadlock

**Review by jmaron99 — APPROVED** [review]
Looks like all discusses modification were made.  Going ahead and approving.

**Review by mgianatagh — APPROVED** [review]

**Review by fredtibbitts — APPROVED** [review]
